[New resource] IntuneMobileAppConfigurationPolicyIos

dannyKBjj opened the pull request:
#### Pull Request (PR) description
<!--
Module adds support for iOS, managed devices, app configuration policies.
#microsoft.graph.iosMobileAppConfiguration

Module is affected by issue #5396. Most settings will not affected. For those that are, replacing ` with " in the M365TenantConfig.ps1 before compiling into a .mof file will fix for now.


-->

#### This Pull Request (PR) fixes the following issues
<!--
None
-->

dannyKBjj committed:
First draft of module

Untested

dannyKBjj committed:
Update MSFT_IntuneMobileAppConfigurationPolicyIOS.psm1

dannyKBjj committed:
Updated .md and .json

dannyKBjj committed:
Update MSFT_IntuneMobileAppConfigurationPolicyIOS.psm1

dannyKBjj committed:
Export Done

Export appears to be working for "settings"

dannyKBjj committed:
Tested OK

Module is affected by issue #5396. Replace ` with " in M365TenantConfig.ps1 is your configuration key values contain double quotes. Otherwise, it should work as expected.

ykuijs commented:
Same comment as for your other PR:
All of our resources require unit tests (which are also completing successfully) and example files, for example:
- Unit tests: https://github.com/microsoft/Microsoft365DSC/blob/Dev/Tests/Unit/Microsoft365DSC/Microsoft365DSC.IntuneDeviceCompliancePolicyiOs.Tests.ps1
- Examples: https://github.com/microsoft/Microsoft365DSC/tree/Dev/Modules/Microsoft365DSC/Examples/Resources/IntuneDeviceCompliancePolicyiOs

Also we need an entry of this new resource in the changelog: https://github.com/microsoft/Microsoft365DSC/blob/Dev/CHANGELOG.md

Could you please add these items to the PR?

dannyKBjj committed:
added tests.ps1

dannyKBjj committed:
added examples

create,update and remove examples added

dannyKBjj commented:
Added unit test, examples and updated change log. Hopefully I did it right?!

ykuijs commented:
Do not see the updated changelog yet. Could you please check?

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneMobileAppConfigurationPolicyIOS/settings.json:
```diff
@@ -0,0 +1,32 @@
+    "description": "This resource configures an Intune Managed Device Mobile App Configuration Policy for iOS Device.",
+    "permissions": {
+        "graph": {
+            "delegated": {
+                "read": [
+                    {
+                        "name": "DeviceManagementApps.Read.All"
+                    }
+                ],
+                "update": [
+                    {
+                        "name": "DeviceManagementApps.ReadWrite.All"
+                    }
+                ]
+            },
+            "application": {
+                "read": [
+                    {
+                        "name": "DeviceManagementApps.Read.All"
+                    }
+                ],
+                "update": [
+                    {
+                        "name": "DeviceManagementApps.ReadWrite.All"
+                    }
+                ]
+            }
+        }
```
Because we're dealing with Intune assignments, an additional entry for `Group.Read.All` is required. You can take at the other Intune resources.

dannyKBjj replied:
should be resolved in latest pull request

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneMobileAppConfigurationPolicyIOS/MSFT_IntuneMobileAppConfigurationPolicyIOS.psm1:
There are many blank linkes in there. Please remove them for better readability.

dannyKBjj replied:
Ok thanks will sort it out over the weekend

dannyKBjj replied:
should be resolved in latest pull request

dannyKBjj committed:
Requested changes made

 Group.Read.All permission added to settings.json

Excessive number of blank lines reduced for readability.

dannyKBjj committed:
Update CHANGELOG.md

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneMobileAppConfigurationPolicyIOS/MSFT_IntuneMobileAppConfigurationPolicyIOS.schema.mof:
Some blank lines are left over here as well. Can you please clean them up a bit?

dannyKBjj replied:
Unnecessary blank lines removed

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneMobileAppConfigurationPolicyIOS/MSFT_IntuneMobileAppConfigurationPolicyIOS.psm1:
```diff
@@ -0,0 +1,770 @@
+
+        if ($null -eq $getValue)
+        {
+            Write-Verbose -Message "Nothing with id {$id} was found"
```
```suggestion
            Write-Verbose -Message "No Intune Mobile App Configuration Policy for iOS with Id {$id} was found"
```
Normally, we want those verbose messages to be clear and concise, directly indicating what resource we're working with since the verbose output might be *very* verbose.

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneMobileAppConfigurationPolicyIOS/MSFT_IntuneMobileAppConfigurationPolicyIOS.psm1:
```diff
@@ -0,0 +1,770 @@
+            return $nullResult
+        }
+
+        Write-Verbose -Message "Found something with id {$id}"
```
Same here.
```suggestion
        Write-Verbose -Message "An Intune Mobile App Configuration Policy for iOS with Id {$id} and DisplayName {$DisplayName} was found"
```

FabienTschanz reviewed (commented):
@dannyKBjj You don't need a separate pull request just for the changelog, adding it here (just as you did) is fine. Although you can omit the MSFT_ prefix in front of the resource name and place it in order 😃

dannyKBjj committed:
Update Modules/Microsoft365DSC/DSCResources/MSFT_IntuneMobileAppConfigurationPolicyIOS/MSFT_IntuneMobileAppConfigurationPolicyIOS.psm1

Co-authored-by: FabienTschanz <71251572+FabienTschanz@users.noreply.github.com>

dannyKBjj committed:
Update Modules/Microsoft365DSC/DSCResources/MSFT_IntuneMobileAppConfigurationPolicyIOS/MSFT_IntuneMobileAppConfigurationPolicyIOS.psm1

Co-authored-by: FabienTschanz <71251572+FabienTschanz@users.noreply.github.com>

dannyKBjj committed:
Removed blank lines from schema.mof

dannyKBjj committed:
Merge branch 'MSFT_IntuneMobileAppConfigurationPolicyIOS' of https://github.com/dannyKBjj/Microsoft365DSC into MSFT_IntuneMobileAppConfigurationPolicyIOS

ykuijs commented on CHANGELOG.md:
```diff
@@ -23,6 +23,8 @@
 * DEPENDENCIES
   * Updated DSCParser to version 2.0.0.12.
   * Updated MSCloudLoginAssistant to version 1.1.28.
+* MSFT_IntuneMobileAppConfigurationPolicyIOS
```
Please remove the MSFT_ prefix and then place this entry in alphabetical order in the list

dannyKBjj committed:
Fixed issues raised on previous pull request

dannyKBjj commented:
Fixed 2-Update.ps1 example. Should now pass unit tests.

dannyKBjj commented:
Unsure how best to resolve CHANGELOG.md conflict?

ykuijs committed:
Merge branch 'Dev' into MSFT_IntuneMobileAppConfigurationPolicyIOS

ykuijs reviewed (approved):
LGTM

dannyKBjj commented:
Hi, see issue #5587  as I believe this module fixes the issue. Have created pull request to re-instate my module and remove the one that this duplicated.

Android apps are handled by IntuneAndroidManagedStoreAppConfiguration , which for whatever reason did not get pulled.

Issue #5587 seems to be caused precisely by the logic the combined resource uses to determine the application assignment. The separated resources do not have this problem.
